x-pack/metricbeat/module/containerd/cpu/metricset.go

```diff
@@ -0,0 +1,243 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package cpu
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/v7/x-pack/metricbeat/module/containerd"
+
+	"github.com/elastic/beats/v7/metricbeat/module/kubernetes/util"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/helper/prometheus"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+// Metricset for apiserver is a prometheus based metricset
+type metricset struct {
+	mb.BaseMetricSet
+	prometheusClient           prometheus.Prometheus
+	prometheusMappings         *prometheus.MetricsMapping
+	calcPct                    bool
+	preSystemCpuUsage          float64
+	preContainerCpuTotalUsage  map[string]float64
+	preContainerCpuKernelUsage map[string]float64
+	preContainerCpuUserUsage   map[string]float64
+}
+
+var _ mb.ReportingMetricSetV2Error = (*metricset)(nil)
+
+// getMetricsetFactory as required by` mb.Registry.MustAddMetricSet`
+func getMetricsetFactory(prometheusMappings *prometheus.MetricsMapping) mb.MetricSetFactory {
+	return func(base mb.BaseMetricSet) (mb.MetricSet, error) {
+		pc, err := prometheus.NewPrometheusClient(base)
+		if err != nil {
+			return nil, err
+		}
+		config := containerd.DefaultConfig()
+		if err := base.Module().UnpackConfig(&config); err != nil {
+			return nil, err
+		}
+		return &metricset{
+			BaseMetricSet:              base,
+			prometheusClient:           pc,
+			prometheusMappings:         prometheusMappings,
+			calcPct:                    config.CalculatePct,
+			preSystemCpuUsage:          0.0,
+			preContainerCpuTotalUsage:  map[string]float64{},
+			preContainerCpuKernelUsage: map[string]float64{},
+			preContainerCpuUserUsage:   map[string]float64{},
+		}, nil
+	}
+}
+
+// Fetch gathers information from the containerd and reports events with this information.
+func (m *metricset) Fetch(reporter mb.ReporterV2) error {
+	events, err := m.prometheusClient.GetProcessedMetrics(m.prometheusMappings)
+	if err != nil {
+		return errors.Wrap(err, "error getting metrics")
+	}
+
+	var systemTotalNs int64
+	perContainerCpus := make(map[string]int)
+
+	if m.calcPct {
+		for _, event := range events {
+			systemTotalSeconds, err := event.GetValue("system.total")
+			if err == nil {
+				systemTotalNs = systemTotalSeconds.(int64) * 1e9
+			}
+			if _, err = event.GetValue("cpu"); err == nil {
+				// calculate cpus used by each container
+				setContCpus(event, perContainerCpus)
+			}
+		}
+	}
+
+	for _, event := range events {
+		// Discard event containing system.total
+		if _, err := event.GetValue("system.total"); err == nil {
+			continue
+		}
+		// setting ECS container.id
+		containerFields := common.MapStr{}
+		var cID string
+		if containerID, ok := event["id"]; ok {
+			cID = (containerID).(string)
+			containerFields.Put("id", cID)
+			event.Delete("id")
+		}
+		e, err := util.CreateEvent(event, "containerd.cpu")
```

Comment:
If we're gonna borrow functions from other modules, we should move that code outside the module to somewhere generic.